src/object.d

```diff
@@ -3241,6 +3241,226 @@ template RTInfo(T)
+                    return -1;
+                }
+            }
+        }
+        return c < 0 ? -1 : c > 0 ? 1 : 0;
+    }
+
+    // floating point types
+    static if ((is(U1 == float) && is(U2 == float))
+        || (is(U1 == double)   && is(U2 == double))
+        || (is(U1 == real)     && is(U2 == real))
+        || (is(U1 == ifloat)   && is(U2 == ifloat))
+        || (is(U1 == idouble)  && is(U2 == idouble))
+        || (is(U1 == ireal)    && is(U2 == ireal))
+        || (is(U1 == cfloat)   && is(U2 == cfloat))
+        || (is(U1 == cdouble)  && is(U2 == cdouble))
+        || (is(U1 == creal)    && is(U2 == creal)))
+    {
+        template Floating(T)
+        if (is(T == cfloat) || is(T == cdouble) || is(T == creal))
+        {
+          pure nothrow @safe:
+            int compare(T f1, T f2)
+            {
+                int result;
+
+                if (f1.re < f2.re)
+                    result = -1;
+                else if (f1.re > f2.re)
+                    result = 1;
+                else if (f1.im < f2.im)
+                    result = -1;
+                else if (f1.im > f2.im)
+                    result = 1;
+                else
+                    result = 0;
+                return result;
+            }
+        }
+
+        template Floating(T)
+        if (is(T == float) || is(T == double) || is(T == real))
+        {
+          pure nothrow @safe:
+            int compare(T d1, T d2)
+            {
+                if (d1 != d1 || d2 != d2) // if either are NaN
+                {
+                    if (d1 != d1)
+                    {
+                        if (d2 != d2)
+                            return 0;
+                        return -1;
+                    }
+                    return 1;
+                }
+                return (d1 == d2) ? 0 : ((d1 < d2) ? -1 : 1);
+            }
+        }
+
+        template Array(T)
+        if (is(T ==  float) || is(T ==  double) || is(T ==  real) ||
+            is(T == cfloat) || is(T == cdouble) || is(T == creal))
+        {
+          pure nothrow @safe:
+            int compare(T[] s1, T[] s2)
+            {
+                size_t len = s1.length;
+                if (s2.length < len)
+                    len = s2.length;
+                for (size_t u = 0; u < len; u++)
+                {
+                    if (int c = Floating!T.compare(s1[u], s2[u]))
+                        return c;
+                }
+                if (s1.length < s2.length)
+                    return -1;
+                else if (s1.length > s2.length)
+                    return 1;
+                return 0;
+            }
+        }
+
+        static if (is(U1 == ifloat)) alias F = float;
+        else static if (is(U1 == idouble)) alias F = double;
+        else static if (is(U1 == ireal)) alias F = real;
+        else alias F = U1;
+
+        return () @trusted { return Array!F.compare(cast(F[])s1, cast(F[])s2); }();
+    }
+
+    // integral, char types
```

Comment:
@WalterBright thx!

1. Calling `dstrcmp` didn't work in this approach because it was invoked during compilation. So we wanted to have something simple that passes muster. At this point @somzzz will look into special casing `dstrcmp` with `if (__ctfe)`.
2. Fixing (1) will also fix (2) - the template becomes an inlined one-liner.
3. using `s1.ptr[u]` has a different issue - it's not deducible as @safe. I've explained @somzzz to use the idiom for generic trusted code.

>For example, float, ifloat, const(float), immutable(float), const(ifloat), immutable(ifloat) should all be forwarded to one non-template function.

That is the case already.

>Next step, sure, but I am concerned that the next step doesn't get taken soon enough and we release a version that has substantially degraded performance. There's no way the manual loop will come close to memcmp().

There actually is if the size is short. On a larger note, this is a long-haul project not a drive-by refactoring so no need to worry.